src/gapstate.h

```diff
@@ -192,21 +194,45 @@ extern GAPState MainGAPState;
 
 #endif
 
-void InitMainGAPState(void);
+/* this is potentially confusing, because MODULE_STATE is supposed to be
+  *local* to the module, so MODULE_STATE will refer to a different bit of
+   state depending which file it is used in */
+#define MODULE_STATE(x) (((ModuleState *)(&STATE(StateSlots)[StateHandle]))->x)
+
+
+// Offset into StateSlots
+typedef Int ModuleStateHandle;
+typedef void (*ModuleConstructor)(ModuleStateHandle handle);
+typedef void (*ModuleDestructor)();
+
+typedef struct {
+    UInt size;
+    Int offset;
+    ModuleConstructor constructor;
+    ModuleDestructor destructor;
+} StateHandler;
+
+/*
+ * Register global state for a module.
+ *
+ * size - how much memory (in bytes) the module would like to reserve in
+ *        GAPState.StateSlots
+ * constructor and destructor - function pointers. These  functions are called
+ *        whenever a module is instantiated or freed, respectively. Currently
+ *        in GAP this only happens once at startup (and exit). In HPC-GAP it
+ *        happens for every thread creation and destruction.
+ *        Constructor and destructor receive the offset (in bytes) into
```

Comment:
This doesn't match the code right now.

Also, if we document this to be an offset, why call it a "handle" elsewhere? Calling it a handle makes sense to me if we want to keep it fully opaque (there certainly is a valid point to be made in favor of that approach). But then there should be a helper function which takes a "handle" and return a pointer to the actual data (for the current thread). In a sense, `MODULE_STATE` does that, though I'd really add a separate such function, and implement `MODULE_STATE` in terms of that; because then code which wants to bypass `MODULE_STATE` can still correctly use the "handle".

Or we decide to keep it simple, and document this as an "offset" (as this comment does); but then I'd call it an "offset", not a "handle", as that'd be much more descriptive.